proposals/v3/api.md

```diff
@@ -0,0 +1,1093 @@
+the list of readings for the device, instead of a collection of 1 `device_id` keyed to this
+device reading info. 
+
+> **Question**: We have a bunch of different response formats for reads - "read", "read device", 
+> and "read cached". They all convey the same data, just structured a bit differently to make it
+> convenient for each of their use cases (bulk read, single read, streamed read -- respectively).
+> Would it make sense to try and unify the response data structures into a more uniform response
+> or just accept that since the endpoints are catered towards different use cases, the format of
+> the data they return will vary?
+
+```json
+[
+  {
+    "kind": "temperature",
+    "type": "temperature",
+    "value": 20.3,
+    "timestamp": "2018-02-01T13:47:40.395939895Z",
+    "info": "",
+    "unit": {
+      "symbol": "C",
+      "name": "degrees celsius"
+    },
+    "context": {
+      "host": "127.0.0.1",
+      "sample_rate": 8
+    }
+  }
+]
+```
+
+##### Error (500, 404)
+See: [Errors](#errors)
+
+
+---
+
+### Read Cached
+```
+GET http://HOST:5000/synse/v3/readcached
+```
+
+Stream reading data from the registered plugins.
+
+All plugins have the capability of caching their readings locally in order to maintain
+a higher resolution of state beyond the poll frequency which Synse Server may request at.
+This is particularly useful for push-based plugins, where we would lose the pushed
+reading if it were not cached.
+
+At the plugin level, caching read data can be enabled, but is disabled by default. Even if
+disabled, this route will still return data for every device that supports reading on each
+of the configured plugins. When read caching is disabled, this will just return a dump of
+the current reading state that is maintained by the plugin.
+
+#### Query Parameters
+
+| Key | Description |
+| --- | ----------- |
+| start | An RFC3339 or RFC3339Nano formatted timestamp which specifies a starting bound on the cache data to return. If no timestamp is specified, there will not be a starting bound. |
+| end | An RFC3339 or RFC3339Nano formatted timestamp which specifies an ending bound on the cache data to return. If no timestamp is specified, there will not be an ending bound. |
+
+#### Response
+
+##### OK (200)
+The response will be streamed JSON. One block of the streamed JSON will appear as follows:
+```json
+{
+  "id": "a72cs6519ee675b",
+  "data": [
+    {
+      "kind": "temperature",
+      "type": "temperature",
+      "value": 20.3,
+      "timestamp": "2018-02-01T13:47:40.395939895Z",
+      "info": "",
+      "unit": {
+        "symbol": "C",
+        "name": "degrees celsius"
+      },
+      "context": {
+        "host": "127.0.0.1",
+        "sample_rate": 8
+      }
+    }
+  ]
+}
+```
+
+This is similar to the [read](#read) response, but organized a bit differently to make it
+more suitable for streaming.
+
+##### Error (500, 400)
+See: [Errors](#errors)
+
+
+---
+
+### Write
+```
```

Comment:
NOTE: The write spec is subject to change (and possibly the transaction stuff since it relates to writes). I was sorta on the edge with allowing bulk writes to begin with but figured I'd spec it out and see what others think. @MatthewHink brought up that its probably a bad idea in that it leads down a complicated rabbit hole of partial successes, and probably other unpleasant things. 

I'm still leaving this as-is for now just to gather a bit more feedback and opinion, but will most likely end up taking  bulk write functionality out of the proposal.